File: Week_02/G20190343020335/[242]Valid Anagram.go
//给定两个字符串 s 和 t ，编写一个函数来判断 t 是否是 s 的字母异位词。 
//
// 示例 1: 
//
// 输入: s = "anagram", t = "nagaram"
//输出: true
// 
//
// 示例 2: 
//
// 输入: s = "rat", t = "car"
//输出: false 
//
// 说明: 
//你可以假设字符串只包含小写字母。 
//
// 进阶: 
//如果输入字符串包含 unicode 字符怎么办？你能否调整你的解法来应对这种情况？ 
// Related Topics 排序 哈希表
package main

//leetcode submit region begin(Prohibit modification and deletion)
func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	// 辣鸡golang，毁我青春，字符串排序那么蛋疼，算了
	countMap := make(map[byte]int)	// 考虑到Unicode字符，就不用int26数组来刷时间复杂度了
	for i:= 0; i < len(s); i ++ {
		countMap[s[i]] += 1
	}
	for i:= 0; i < len(t); i ++  {
		if _, ok := countMap[t[i]]; !ok {
			return false
		}
		countMap[t[i]] -= 1
	}
	for _, cnt := range countMap {
		if cnt != 0 {
			return false
		}
	}
	return true
}
//leetcode submit region end(Prohibit modification and deletion)

File: Week_02/G20190343020335/[242]Valid Anagram.py
#给定两个字符串 s 和 t ，编写一个函数来判断 t 是否是 s 的字母异位词。 
#
# 示例 1: 
#
# 输入: s = "anagram", t = "nagaram"
#输出: true
# 
#
# 示例 2: 
#
# 输入: s = "rat", t = "car"
#输出: false 
#
# 说明: 
#你可以假设字符串只包含小写字母。 
#
# 进阶: 
#如果输入字符串包含 unicode 字符怎么办？你能否调整你的解法来应对这种情况？ 
# Related Topics 排序 哈希表



#leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # 长度不一致直接false
        if len(s) != len(t):
            return False
        # 方法一：一行解决 O(nlogn)
        # return sorted(s) == sorted(t)

        #方法二：hashmap计数
        count_map = {}
        for c in s:
            if count_map.get(c, -1) == -1:
                count_map[c] = 0
            count_map[c] += 1
        for c in t:
            if count_map.get(c, -1) == -1:
                return False
            count_map[c] -= 1
        for _, v in count_map:
            if v != 0:
                return False
        return True


#leetcode submit region end(Prohibit modification and deletion)
